File: 19_数学/计算几何/基础/直线/447. 回旋镖的数量.js
/**
 * @param {number[][]} points
 * @return {number}   返回平面上所有回旋镖的数量
 * 我们可以枚举每个 points[i]，将其当作 V 型的拐点。
 */
function numberOfBoomerangs(points) {
  let res = 0
  for (const p1 of points) {
    const counter = new Map()
    for (const p2 of points) {
      const dis = (p1[0] - p2[0]) * (p1[0] - p2[0]) + (p1[1] - p2[1]) * (p1[1] - p2[1])
      counter.set(dis, (counter.get(dis) || 0) + 1)
    }
    for (const [_, m] of counter.entries()) {
      res += m * (m - 1)
    }
  }
  return res
}

File: 5_map/447. 回旋镖的数量.py
from typing import List
from collections import Counter
from math import dist


# 给定平面上 n 对 互不相同 的点 points
# 回旋镖 是由点 (i, j, k) 表示的元组 ，
# 其中 i 和 j 之间的距离和 i 和 k 之间的欧式距离相等（需要考虑元组的顺序）。
# 返回平面上所有回旋镖的数量。


class Solution:
    def numberOfBoomerangs(self, points: List[List[int]]) -> int:
        res = 0
        for p2 in points:
            counter = Counter()
            for p1 in points:
                if p1 == p2:
                    continue
                counter[dist(p1, p2)] += 1
            res += sum(v * (v - 1) for v in counter.values())
        return res


print(Solution().numberOfBoomerangs([[0, 0], [1, 0], [2, 0]]))
print(Solution().numberOfBoomerangs([[0, 0], [1, 0], [-1, 0], [0, 1], [0, -1]]))

